core/shared/src/main/scala/fs2/retry.scala

```diff
@@ -0,0 +1,57 @@
+package fs2
+
+import cats.effect.Async
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.FiniteDuration
+import scala.util.control.NonFatal
+
+object retry {
+  /**
+   * Retries `fa` on failure, returning a single element
+   * stream with the result of `fa` as soon as it succeeds.
+   *
+   * @param delay Duration of delay before the first retry
+   *
+   * @param nextDelay Applied to the previous delay to compute the
+   *                  next, e.g. to implement exponential backoff
+   *
+   * @param maxRetries Number of attempts before failing with the
+   *                   latest error, if `fa` never succeeds
+   *
+   * @param retriable Function to determine whether a failure is
+   *                  retriable or not, defaults to retry every
+   *                  `NonFatal`. A failed stream is immediately
+   *                  returned when a non-retriable failure is
+   *                  encountered
+   */
+  def retry[F[_], A](fa: F[A],
+                     delay: FiniteDuration,
```

Comment:
Let's set  aside the `Stream` vs `F` debate for a second: it's a very small implementation change), and I do hear yours and @pchlupacek arguments about resource safety, so definitely open to change that bit.

As for the API, my idea was to provide a very simple, but powerful and compositional method  for people to build upon (that would be `retries/attempts`): I 100% agree that ability to work with standard combinators is paramount, and imo compositionality is the best feature of fs2 so I wouldn't want to lose that.

As for `retry`, I also wanted to provide people with an almost drop-in replacement to scalaz.Task.retry, hence the specific API. Would it be possible to build your `retryWithDelay` on top of attempts? I haven't tried, but my gut feeling is that it would.